// @speckle/shared/dist/esm/onboarding/helpers/index.js
export var OnboardingRole;
(function (OnboardingRole) {
    OnboardingRole["ComputationalDesign"] = "computational-design";
    OnboardingRole["BIM"] = "bim";
    OnboardingRole["ArchitecturePlanning"] = "architecture-planning";
    OnboardingRole["EngineeringAEC"] = "engineering-aec";
    OnboardingRole["EngineeringSoftware"] = "engineering-software";
    OnboardingRole["Education"] = "education";
    OnboardingRole["Management"] = "management";
    OnboardingRole["Other"] = "other";
})(OnboardingRole || (OnboardingRole = {}));
export var OnboardingPlan;
(function (OnboardingPlan) {
    OnboardingPlan["Exploring"] = "exploring";
    OnboardingPlan["DataExchange"] = "data-exchange";
    OnboardingPlan["Analytics"] = "analytics";
    OnboardingPlan["Collaboration"] = "collaboration";
    OnboardingPlan["DataWarehouse"] = "data-warehouse";
    OnboardingPlan["Development"] = "development";
    OnboardingPlan["Automation"] = "automation";
})(OnboardingPlan || (OnboardingPlan = {}));
export var OnboardingSource;
(function (OnboardingSource) {
    OnboardingSource["SocialMedia"] = "social-media";
    OnboardingSource["Search"] = "internet-search";
    OnboardingSource["Referral"] = "friend-or-colleague";
    OnboardingSource["Event"] = "event-conference";
    OnboardingSource["Education"] = "university-course";
    OnboardingSource["Other"] = "other";
})(OnboardingSource || (OnboardingSource = {}));
