// true-myth/dist/cjs/index.cjs
"use strict";
/**
  This is just here to re-export {@link Maybe}, {@link Result}, {@link Unit},
  and {@link Toolbelt}  – that is, to provide a root-level entry amenable to use
  with Node with TypeScript versions before 4.7 and its ESM support, or for
  convenience.

  @packageDocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Toolbelt = exports.Task = exports.Unit = exports.ResultNS = exports.Result = exports.MaybeNS = exports.Maybe = void 0;
var maybe_js_1 = require("./maybe.cjs");
Object.defineProperty(exports, "Maybe", { enumerable: true, get: function () { return maybe_js_1.default; } });
exports.MaybeNS = require("./maybe.cjs");
var result_js_1 = require("./result.cjs");
Object.defineProperty(exports, "Result", { enumerable: true, get: function () { return result_js_1.default; } });
exports.ResultNS = require("./result.cjs");
var unit_js_1 = require("./unit.cjs");
Object.defineProperty(exports, "Unit", { enumerable: true, get: function () { return unit_js_1.default; } });
var task_js_1 = require("./task.cjs");
Object.defineProperty(exports, "Task", { enumerable: true, get: function () { return task_js_1.default; } });
exports.Toolbelt = require("./toolbelt.cjs");
