// @speckle/shared/dist/commonjs/authz/fragments/dashboards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensureDashboardProjectsReadAccess = exports.ensureWorkspaceDashboardsFeatureAccessFragment = exports.ensureDashboardsEnabledFragment = void 0;
const result_1 = require("true-myth/result");
const authErrors_js_1 = require("../domain/authErrors.js");
const index_js_1 = require("../../workspaces/index.js");
const projects_js_1 = require("./projects.js");
const ensureDashboardsEnabledFragment = (loaders) => async () => {
    const env = await loaders.getEnv();
    if (!env.FF_DASHBOARDS_MODULE_ENABLED)
        return (0, result_1.err)(new authErrors_js_1.DashboardsNotEnabledError());
    return (0, result_1.ok)();
};
exports.ensureDashboardsEnabledFragment = ensureDashboardsEnabledFragment;
const ensureWorkspaceDashboardsFeatureAccessFragment = (loaders) => async ({ workspaceId }) => {
    const plan = await loaders.getWorkspacePlan({ workspaceId });
    if (!plan)
        return (0, result_1.err)(new authErrors_js_1.WorkspacePlanNoFeatureAccessError());
    const isFlagOn = (0, index_js_1.isWorkspaceFeatureFlagOn)({
        workspaceFeatureFlags: plan.featureFlags,
        feature: index_js_1.WorkspaceFeatureFlags.dashboards
    });
    if (!isFlagOn)
        return (0, result_1.err)(new authErrors_js_1.WorkspacePlanNoFeatureAccessError());
    return (0, result_1.ok)();
};
exports.ensureWorkspaceDashboardsFeatureAccessFragment = ensureWorkspaceDashboardsFeatureAccessFragment;
const ensureDashboardProjectsReadAccess = (loaders) => async ({ userId, dashboardId }) => {
    const dashboard = await loaders.getDashboard({ dashboardId });
    if (!dashboard)
        return (0, result_1.err)(new authErrors_js_1.DashboardNotFoundError());
    const allProjectResults = await Promise.all(dashboard.projectIds.map(async (projectId) => {
        return [
            projectId,
            await (0, projects_js_1.ensureMinimumProjectRoleFragment)(loaders)({ projectId, userId })
        ];
    }));
    const projectAccessErrors = allProjectResults.filter(([, e]) => e.isErr);
    return projectAccessErrors.length
        ? (0, result_1.err)(new authErrors_js_1.DashboardProjectsNotEnoughPermissionsError({
            payload: {
                projectIds: projectAccessErrors.map(([projectId]) => projectId)
            }
        }))
        : (0, result_1.ok)();
};
exports.ensureDashboardProjectsReadAccess = ensureDashboardProjectsReadAccess;
