// @speckle/shared/dist/commonjs/authz/policies/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.authPoliciesFactory = void 0;
const canCreateWorkspaceProject_js_1 = require("./workspace/canCreateWorkspaceProject.js");
const canRead_js_1 = require("./project/canRead.js");
const canCreate_js_1 = require("./project/model/canCreate.js");
const canMoveToWorkspace_js_1 = require("./project/canMoveToWorkspace.js");
const canCreatePersonal_js_1 = require("./project/canCreatePersonal.js");
const canUpdate_js_1 = require("./project/canUpdate.js");
const canReadSettings_js_1 = require("./project/canReadSettings.js");
const canReadWebhooks_js_1 = require("./project/canReadWebhooks.js");
const canUpdateAllowPublicComments_js_1 = require("./project/canUpdateAllowPublicComments.js");
const canLeave_js_1 = require("./project/canLeave.js");
const canInvite_js_1 = require("./workspace/canInvite.js");
const canInvite_js_2 = require("./project/canInvite.js");
const canBroadcastActivity_js_1 = require("./project/canBroadcastActivity.js");
const canCreate_js_2 = require("./project/comment/canCreate.js");
const canArchive_js_1 = require("./project/comment/canArchive.js");
const canEdit_js_1 = require("./project/comment/canEdit.js");
const canUpdate_js_2 = require("./project/model/canUpdate.js");
const canDelete_js_1 = require("./project/model/canDelete.js");
const canCreate_js_3 = require("./project/version/canCreate.js");
const canUpdate_js_3 = require("./project/version/canUpdate.js");
const canRequestRender_js_1 = require("./project/version/canRequestRender.js");
const canCreate_js_4 = require("./project/automation/canCreate.js");
const canUpdate_js_4 = require("./project/automation/canUpdate.js");
const canRead_js_2 = require("./project/automation/canRead.js");
const canReceiveProjectsUpdatedMessage_js_1 = require("./workspace/canReceiveProjectsUpdatedMessage.js");
const canDelete_js_2 = require("./project/canDelete.js");
const canDelete_js_3 = require("./project/automation/canDelete.js");
const canPublish_js_1 = require("./project/canPublish.js");
const canLoad_js_1 = require("./project/canLoad.js");
const canReadMemberEmail_js_1 = require("./workspace/canReadMemberEmail.js");
const canCreateWorkspace_js_1 = require("./workspace/canCreateWorkspace.js");
const canUseWorkspacePlanFeature_js_1 = require("./workspace/canUseWorkspacePlanFeature.js");
const canEditFunction_js_1 = require("./automate/function/canEditFunction.js");
const canUpdateEmbedTokens_js_1 = require("./project/canUpdateEmbedTokens.js");
const canReadAccIntegrationSettings_js_1 = require("./project/canReadAccIntegrationSettings.js");
const canCreate_js_5 = require("./project/savedViews/canCreate.js");
const canUpdate_js_5 = require("./project/savedViews/canUpdate.js");
const canUpdateGroup_js_1 = require("./project/savedViews/canUpdateGroup.js");
const canRead_js_3 = require("./project/savedViews/canRead.js");
const canListDashboards_js_1 = require("./workspace/canListDashboards.js");
const canDelete_js_4 = require("./dashboard/canDelete.js");
const canCreateDashboards_js_1 = require("./workspace/canCreateDashboards.js");
const canCreateToken_js_1 = require("./dashboard/canCreateToken.js");
const canEdit_js_2 = require("./dashboard/canEdit.js");
const canRead_js_4 = require("./dashboard/canRead.js");
const canMove_js_1 = require("./project/savedViews/canMove.js");
const canEditTitle_js_1 = require("./project/savedViews/canEditTitle.js");
const canEditDescription_js_1 = require("./project/savedViews/canEditDescription.js");
const canCreateSavedViewGroupToken_js_1 = require("./project/savedViews/canCreateSavedViewGroupToken.js");
const canSetAsHomeView_js_1 = require("./project/savedViews/canSetAsHomeView.js");
const authPoliciesFactory = (loaders) => ({
    automate: {
        function: {
            canRegenerateToken: (0, canEditFunction_js_1.canEditFunctionPolicy)(loaders)
        }
    },
    dashboard: {
        canCreateToken: (0, canCreateToken_js_1.canCreateDashboardTokenPolicy)(loaders),
        canDelete: (0, canDelete_js_4.canDeleteDashboardPolicy)(loaders),
        canEdit: (0, canEdit_js_2.canEditDashboardPolicy)(loaders),
        canRead: (0, canRead_js_4.canReadDashboardPolicy)(loaders)
    },
    project: {
        automation: {
            canCreate: (0, canCreate_js_4.canCreateAutomationPolicy)(loaders),
            canRead: (0, canRead_js_2.canReadAutomationPolicy)(loaders),
            canUpdate: (0, canUpdate_js_4.canUpdateAutomationPolicy)(loaders),
            canDelete: (0, canDelete_js_3.canDeleteAutomationPolicy)(loaders)
        },
        model: {
            canCreate: (0, canCreate_js_1.canCreateModelPolicy)(loaders),
            canUpdate: (0, canUpdate_js_2.canUpdateModelPolicy)(loaders),
            canDelete: (0, canDelete_js_1.canDeleteModelPolicy)(loaders)
        },
        comment: {
            canCreate: (0, canCreate_js_2.canCreateProjectCommentPolicy)(loaders),
            canArchive: (0, canArchive_js_1.canArchiveProjectCommentPolicy)(loaders),
            canEdit: (0, canEdit_js_1.canEditProjectCommentPolicy)(loaders)
        },
        version: {
            canCreate: (0, canCreate_js_3.canCreateProjectVersionPolicy)(loaders),
            canUpdate: (0, canUpdate_js_3.canUpdateProjectVersionPolicy)(loaders),
            canReceive: (0, canLoad_js_1.canLoadPolicy)(loaders),
            canRequestRender: (0, canRequestRender_js_1.canRequestProjectVersionRenderPolicy)(loaders)
        },
        savedViews: {
            canCreate: (0, canCreate_js_5.canCreateSavedViewPolicy)(loaders),
            canUpdate: (0, canUpdate_js_5.canUpdateSavedViewPolicy)(loaders),
            canUpdateGroup: (0, canUpdateGroup_js_1.canUpdateSavedViewGroupPolicy)(loaders),
            canCreateToken: (0, canCreateSavedViewGroupToken_js_1.canCreateSavedViewGroupTokenPolicy)(loaders),
            canRead: (0, canRead_js_3.canReadSavedViewPolicy)(loaders),
            canMove: (0, canMove_js_1.canMoveSavedViewPolicy)(loaders),
            canEditTitle: (0, canEditTitle_js_1.canEditSavedViewTitlePolicy)(loaders),
            canEditDescription: (0, canEditDescription_js_1.canEditSavedViewDescriptionPolicy)(loaders),
            canSetAsHomeView: (0, canSetAsHomeView_js_1.canSetSavedViewAsHomeViewPolicy)(loaders)
        },
        canBroadcastActivity: (0, canBroadcastActivity_js_1.canBroadcastProjectActivityPolicy)(loaders),
        canRead: (0, canRead_js_1.canReadProjectPolicy)(loaders),
        canMoveToWorkspace: (0, canMoveToWorkspace_js_1.canMoveToWorkspacePolicy)(loaders),
        canCreatePersonal: (0, canCreatePersonal_js_1.canCreatePersonalProjectPolicy)(loaders),
        canUpdate: (0, canUpdate_js_1.canUpdateProjectPolicy)(loaders),
        canDelete: (0, canDelete_js_2.canDeleteProjectPolicy)(loaders),
        canUpdateAllowPublicComments: (0, canUpdateAllowPublicComments_js_1.canUpdateProjectAllowPublicCommentsPolicy)(loaders),
        canReadSettings: (0, canReadSettings_js_1.canReadProjectSettingsPolicy)(loaders),
        canReadWebhooks: (0, canReadWebhooks_js_1.canReadProjectWebhooksPolicy)(loaders),
        canLeave: (0, canLeave_js_1.canLeaveProjectPolicy)(loaders),
        canInvite: (0, canInvite_js_2.canInviteToProjectPolicy)(loaders),
        canPublish: (0, canPublish_js_1.canPublishPolicy)(loaders),
        canLoad: (0, canLoad_js_1.canLoadPolicy)(loaders),
        canReadEmbedTokens: (0, canUpdateEmbedTokens_js_1.canUpdateEmbedTokensPolicy)(loaders),
        canUpdateEmbedTokens: (0, canUpdateEmbedTokens_js_1.canUpdateEmbedTokensPolicy)(loaders),
        canReadAccIntegrationSettings: (0, canReadAccIntegrationSettings_js_1.canReadAccIntegrationSettingsPolicy)(loaders),
        canUpdateAccIntegrationSettings: (0, canReadAccIntegrationSettings_js_1.canReadAccIntegrationSettingsPolicy)(loaders)
    },
    workspace: {
        canCreateProject: (0, canCreateWorkspaceProject_js_1.canCreateWorkspaceProjectPolicy)(loaders),
        canInvite: (0, canInvite_js_1.canInviteToWorkspacePolicy)(loaders),
        canReceiveProjectsUpdatedMessage: (0, canReceiveProjectsUpdatedMessage_js_1.canReceiveWorkspaceProjectsUpdatedMessagePolicy)(loaders),
        canUseWorkspacePlanFeature: (0, canUseWorkspacePlanFeature_js_1.canUseWorkspacePlanFeature)(loaders),
        canReadMemberEmail: (0, canReadMemberEmail_js_1.canReadMemberEmailPolicy)(loaders),
        canCreateWorkspace: (0, canCreateWorkspace_js_1.canCreateWorkspacePolicy)(loaders),
        canCreateDashboards: (0, canCreateDashboards_js_1.canCreateDashboardsPolicy)(loaders),
        canListDashboards: (0, canListDashboards_js_1.canListDashboardsPolicy)(loaders)
    }
});
exports.authPoliciesFactory = authPoliciesFactory;
